src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -1978,66 +1978,80 @@ private void checkpoint() {
         FlowNodeStorage storage = getStorage();
         if (storage != null) {
             try { // Node storage must be flushed first so program can be restored
                 storage.flush();
             } catch (IOException ioe) {
                 persistOk=false;
-                LOGGER.log(Level.WARNING, "Error persisting FlowNode storage before shutdown", ioe);
+                LOGGER.log(Level.WARNING, "Error persisting FlowNode storage for: " + this, ioe);
             }
 
             // Try to ensure we've saved the appropriate things -- the program is the last stumbling block.
             try {
                 final SettableFuture<Void> myOutcome = SettableFuture.create();
-                LOGGER.log(Level.INFO, "About to try to checkpoint the program for build"+this);
+                LOGGER.log(Level.FINE, "About to try to checkpoint the program for: {0}", this);
                 if (programPromise != null && programPromise.isDone()) {
                     runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
                         @Override
                         public void onSuccess(CpsThreadGroup result) {
                             try {
-                                LOGGER.log(Level.INFO, "Trying to save program before shutdown "+this);
+                                LOGGER.log(Level.FINE, "Trying to save program for: {0}", CpsFlowExecution.this);
                                 result.saveProgramIfPossible(true);
-                                LOGGER.log(Level.INFO, "Finished saving program before shutdown "+this);
+                                LOGGER.log(Level.FINE, "Finished saving program for: {0}", CpsFlowExecution.this);
                                 myOutcome.set(null);
                             } catch (Exception ex) {
-                                LOGGER.log(Level.WARNING, "Error persisting program: "+ex);
+                                // Logged at Level.WARNING when we call `myOutcome.get` and it throws an exception.
                                 myOutcome.setException(ex);
                             }
                         }
 
                         @Override
                         public void onFailure(Throwable t) {
-                            LOGGER.log(Level.WARNING, "Failed trying to save program before shutdown "+this);
+                            // Logged at Level.WARNING when we call `myOutcome.get` and it throws an exception.
                             myOutcome.setException(t);
                         }
                     });
                     myOutcome.get(30, TimeUnit.SECONDS);
-                    LOGGER.log(Level.FINE, "Successfully saved program for "+this);
+                    LOGGER.log(Level.FINE, "Successfully saved program for: {0}", this);
+                } else {
+                    persistOk = false;
+                    LOGGER.log(Level.WARNING, "Unable to persist program because it was never loaded for: {0}", this);
```

Comment:
I don't think this should ever happen, but definitely worth a log message if it does!